AD9545: Add driver and basic project example

jonathanns opened the pull request:
## Pull Request Description

Add support for AD9545 Quad Input, 10-Output, Dual DPLL/IEEE 1588, 1 pps Synchronizer and Jitter Cleaner. Also includes a Basic project example tested in AD-SYNCHRONA14-EBZ.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

jonathanns commented:
This is a WIP in the final stages.
Some dependencies are available on #2268.

danmois commented on drivers/frequency/ad9545/ad9545.h:
```diff
@@ -0,0 +1,826 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef ad9545_H_
```
Caps?

jonathanns replied:
noted

danmois commented on drivers/frequency/ad9545/ad9545.h:
```diff
@@ -0,0 +1,826 @@
+	bool				pll_used;
+	unsigned int			address;
+	struct no_os_clk_desc		*hw;
+	uint8_t			num_parents;
```
Alignment

jonathanns replied:
noted

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+
+	ret = ad9545_read_reg_multiple(dev, AD9545_REF_X_RDIV(addr), (uint8_t *)&regval,
+				       4);
+	if (ret < 0)
```
if (ret) ?

jonathanns replied:
Sure, i can replace.

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+
+static int32_t ad9545_check_id(struct ad9545_dev *dev)
+{
+
```
Extra line

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+static int ad9545_set_r_div(struct ad9545_dev *dev, uint32_t div, int addr)
+{
+	uint32_t regval;
+	uint32_t val;
```
Is val required in this function?

jonathanns replied:
Actually not. removing...

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+
+	dev = (struct ad9545_dev *)no_os_malloc(sizeof(*dev));
+	if (!dev)
+		goto error;
```
What will we return from here?

jonathanns replied:
Actually it should be:
```
if (!dev) 
           return -ENOMEM;
```
if you agree.

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+
+	/* Device settings */
+	ret = ad9545_parse_init(dev, init_param);
+	if (!ret) {
```
Looks somehow different than what we have. I'm not saying that it is not correct, only that I am used with the flow below:
if (ret)
   goto error;

*device = dev; ...

jonathanns replied:
Sure, i agree with you.

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+	for (i = 4; i < 256; i++) {
+		fs = i * fosc;
+
+		if (fs > 2250 && fs < 2415) {
```
Could we have macros here?

jonathanns replied:
noted

danmois commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+				return ret;
+
+			cal_count = 2;
+			break;
```
Is break required here?

jonathanns replied:
In this case it seems redundant with the previous line.

danmois commented on projects/ad9545/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,275 @@
+	if (ret < 0)
+		return ret;
+
+	while(1) {
```
the program could finish with a "Done" message if everything was ok. Do we need the loop here after the device is setup?

jonathanns replied:
Yes, indeed the loop is not necessary, maybe i left there after the tests, i will remove.

jonathanns commented on drivers/frequency/ad9545/ad9545.h:
```diff
@@ -0,0 +1,826 @@
+struct ad9545_dev;
+
+typedef int32_t (*ad9545_reg_read_func)(struct ad9545_dev *dev,
+					uint16_t reg_addr,
```
noted

jonathanns commented on drivers/frequency/ad9545/ad9545.h:
```diff
@@ -0,0 +1,826 @@
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+int32_t ad9545_spi_reg_read(struct ad9545_dev *dev,
```
ok!

jonathanns commented on drivers/frequency/ad9545/ad9545.c:
```diff
@@ -0,0 +1,2434 @@
+	uint64_t regval64;
+	int ret;
+
+	regval64 = 0;
```
ok

jonathanns commented:
Update v1: 
* Fixed nits.
* Added brief descriptions to function headers.
* Replaced constant values with macros.
* Removed unnecessary variables.

danmois commented on projects/ad9545/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,274 @@
+		return ret;
+
+	printf("Done.\n\r");
+	return 0;
```
We should also remove the device descriptor before exiting the program.

jonathanns commented:
Update v2: 
* Fixed the `ad9545_remove` function.
* Added device descriptor removal before exiting program.

danmois reviewed (approved):
lgtm
